### PR TITLE
feat: impl duration datatype and vectors

### PR DESCRIPTION
**Pull request by QuenKar**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

_PLEASE DO NOT LEAVE THIS EMPTY !!!_

Please explain IN DETAIL what the changes are in this PR and why they are needed:

Main change:
1. implement `DurationType` in `src/datatypes/src/types`.
2. implement Datatype, Vector, Scalar, WrapperType traits for `Duration` type, we can use `DurationSecondVector`,`DurationMillisecondVector` etc.
3. implement some conversion between `Duration` and gRPC, JSON, arrow duration arrays, ScalarValue etc.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by killme2008** [comment]
@QuenKar Thanks for your PR. Is this PR ready for review? If so, there are some conflicts that have to be resolved.

**Comment by QuenKar** [comment]
@killme2008 thx, I resolved the conflicts, please take a look.

File: Cargo.toml [hunk 1]
```diff
@@ -77,7 +77,7 @@ datafusion-substrait = { git = "https://github.com/waynexia/arrow-datafusion.git
 derive_builder = "0.12"
 futures = "0.3"
 futures-util = "0.3"
-greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "3489b4742150abe0a769faf1bb60fbb95b061fc8" }
+greptime-proto = { git = "https://github.com/Quenkar/greptime-proto.git", rev = "80d3939ed32564abcde7f8257c65c194d095dad8" }
```
Comment (killme2008) [inline Cargo.toml:80]: Can we use the official branch?

File: src/api/src/helper.rs [hunk 2]
```diff
@@ -404,6 +441,10 @@ pub fn pb_value_to_value_ref(value: &v1::Value) -> ValueRef {
             let interval = Interval::from_month_day_nano(v.months, v.days, v.nanoseconds);
             ValueRef::Interval(interval)
         }
+        ValueData::DurSecondValue(v) => ValueRef::Duration(Duration::new_second(*v)),
```
Comment (killme2008) [inline src/api/src/helper.rs:444]: I think using the complete `Duration` is better than `Dur` prefix. What's your opinion?  @waynexia

File: src/servers/src/postgres/types.rs [hunk 3]
```diff
@@ -103,12 +103,14 @@ pub(super) fn encode_value(value: &Value, builder: &mut DataRowEncoder) -> PgWir
             }
         }
         Value::Interval(v) => builder.encode_field(&PgInterval::from(*v)),
-        Value::List(_) => Err(PgWireError::ApiError(Box::new(Error::Internal {
-            err_msg: format!(
-                "cannot write value {:?} in postgres protocol: unimplemented",
-                &value
-            ),
-        }))),
+        Value::List(_) | Value::Duration(_) => {
```
Comment (killme2008) [inline src/servers/src/postgres/types.rs:106]: Why can't write the duration to pg?

**Review by killme2008 — COMMENTED** [review]
Good job!

**Comment by killme2008** [comment]
@waynexia Please take a look

File: src/api/src/helper.rs [hunk 4]
```diff
@@ -255,6 +270,22 @@ pub fn values_with_capacity(datatype: ColumnDataType, capacity: usize) -> Values
             interval_month_day_nano_values: Vec::with_capacity(capacity),
             ..Default::default()
         },
+        ColumnDataType::DurationSecond => Values {
+            dur_second_values: Vec::with_capacity(capacity),
```
Comment (evenyag) [inline src/api/src/helper.rs:274]: Prefer `duration`
```suggestion
            duration_second_values: Vec::with_capacity(capacity),
```

File: src/datatypes/src/data_type.rs [hunk 5]
```diff
@@ -204,6 +212,15 @@ impl ConcreteDataType {
             _ => None,
         }
     }
+
+    /// Try to cast data type as a [`DurationType`].
+    pub fn as_duration(&self) -> Option<DurationType> {
+        match self {
+            ConcreteDataType::Int64(_) => Some(DurationType::Millisecond(DurationMillisecondType)),
```
Comment (evenyag) [inline src/datatypes/src/data_type.rs:219]: Maybe we don't need to support this
```suggestion
```

File: src/datatypes/src/types/duration_type.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                            ),
+                        })
+                }
+
+                fn cast_value_ref(value: ValueRef) -> crate::Result<Option<Self::Wrapper>> {
+                    match value {
+                        ValueRef::Null => Ok(None),
+                        ValueRef::Int64(v) =>{
+                            Ok(Some([<Duration $unit>]::from(v)))
+                        }
```
Comment (evenyag) [inline src/datatypes/src/types/duration_type.rs:139]: Can we remove this support? @v0y4g3r @waynexia

File: src/sql/src/statements.rs [hunk 7]
```diff
@@ -401,11 +401,14 @@ pub fn concrete_data_type_to_sql_data_type(data_type: &ConcreteDataType) -> Resu
             Some(time_type.precision()),
             TimezoneInfo::None,
         )),
+        ConcreteDataType::Interval(_) => Ok(SqlDataType::Interval),
         ConcreteDataType::Binary(_) => Ok(SqlDataType::Varbinary(None)),
-        ConcreteDataType::Null(_) | ConcreteDataType::List(_) | ConcreteDataType::Dictionary(_) => {
+        ConcreteDataType::Duration(_)
```
Comment (evenyag) [inline src/sql/src/statements.rs:406]: Does sql parser provide a duration type?

File: src/servers/src/postgres/types.rs (hunk 3 above)
Comment (QuenKar) [reply src/servers/src/postgres/types.rs:106]: yep, pg doesn't support for `duration` type.

File: src/sql/src/statements.rs (hunk 7 above)
Comment (QuenKar) [reply src/sql/src/statements.rs:419]: sql parser doesn't provide.

**Comment by QuenKar** [comment]
@evenyag If you have time, please take a look(•̀ᴗ• )

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by evenyag — APPROVED** [review]

**Pull request merged**